src/main/java/com/redhat/devtools/lsp4ij/features/codeBlockProvider/LSPCodeBlockUtils.java

```diff
@@ -0,0 +1,141 @@
+/*******************************************************************************
+ * Copyright (c) 2024 Red Hat, Inc.
+ * Distributed under license by Red Hat, Inc. All rights reserved.
+ * This program is made available under the terms of the
+ * Eclipse Public License v2.0 which accompanies this distribution,
+ * and is available at http://www.eclipse.org/legal/epl-v20.html
+ *
+ * Contributors:
+ * Red Hat, Inc. - initial API and implementation
+ ******************************************************************************/
+
+package com.redhat.devtools.lsp4ij.features.codeBlockProvider;
+
+import com.intellij.ide.highlighter.custom.SyntaxTable;
+import com.intellij.openapi.fileTypes.FileType;
+import com.intellij.openapi.fileTypes.impl.AbstractFileType;
+import com.intellij.psi.PsiFile;
+import com.intellij.psi.util.CachedValueProvider.Result;
+import com.intellij.psi.util.CachedValuesManager;
+import com.intellij.util.containers.ContainerUtil;
+import com.redhat.devtools.lsp4ij.LSPIJTextMateUtils;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.plugins.textmate.psi.TextMateFile;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Utilities for deriving information about code blocks.
+ */
+@ApiStatus.Internal
+public final class LSPCodeBlockUtils {
```

Comment:
I derived another utility class specifically for code block-related utility methods such as `isCodeBlockStart/EndChar()` and `getCodeBlockStart/EndChar()`. This is also what tries to use the file type-specific information to derive supported brace pairs. It works for both TextMate files and plain text files backed by `AbstractFileType` information, but as stated a few times already, the TextMate aspect is currently commented out due to a breaking API change across the supported IDE versions.